Pod/Classes/PINProgressiveImage.m

```diff
@@ -350,59 +301,135 @@ - (CFTimeInterval)estimatedRemainingTime
 - (UIImage *)postProcessImage:(UIImage *)inputImage withProgress:(float)progress
 {
+    UIImage *outputUIImage = nil;
     CGImageRef inputImageRef = CGImageRetain(inputImage.CGImage);
     if (inputImageRef == nil) {
         return nil;
     }
     
-    CIFilter *gaussianFilter = [CIFilter filterWithName:@"CIGaussianBlur"];
-    [gaussianFilter setDefaults];
-    
-    UIImage *outputUIImage = nil;
-    
-    CIContext *processingContext = GPUProcessingContext;
-    if (self.inBackground) {
-        processingContext = CPUProcessingContext;
-    }
-    
-    CGSize maxInput = processingContext.inputImageMaximumSize;
-    CGSize maxOutput = processingContext.outputImageMaximumSize;
+    CGSize maxSize = CGSizeMake(1024, 1024);
     CGSize inputSize = inputImage.size;
-    if (maxInput.height < inputSize.height ||
-        maxInput.width < inputSize.width ||
-        maxOutput.height < inputSize.height ||
-        maxOutput.width < inputSize.width) {
+    if (inputSize.width < 1 ||
+        inputSize.height < 1 ||
+        maxSize.height < inputSize.height ||
+        maxSize.width < inputSize.width) {
         CGImageRelease(inputImageRef);
         return nil;
     }
     
-    if (processingContext && gaussianFilter) {
-        [gaussianFilter setValue:[CIImage imageWithCGImage:inputImageRef]
-                          forKey:kCIInputImageKey];
+    CGFloat radius = MAX(__FLT_EPSILON__, (inputImage.size.width / 25.0) * MAX(0, 1.0 - progress));
+    radius *= inputImage.scale;
+    
+    UIGraphicsBeginImageContextWithOptions(inputSize, YES, inputImage.scale);
+    CGContextRef ctx = UIGraphicsGetCurrentContext();
+    
+    if (ctx) {
+        CGContextScaleCTM(ctx, 1.0, -1.0);
+        CGContextTranslateCTM(ctx, 0, -inputSize.height);
         
-        NSNumber *radius = @((inputImage.size.width / 50.0) * MAX(0, 1.0 - progress));
-        [gaussianFilter setValue:radius
-                          forKey:kCIInputRadiusKey];
+        vImage_Buffer effectInBuffer;
+        vImage_Buffer scratchBuffer;
         
-        CIImage *outputImage = [gaussianFilter outputImage];
-        if (outputImage) {
-            __block CGImageRef outputImageRef = nil;
-            dispatch_sync([[self class] processingQueue], ^{
-                outputImageRef = [processingContext createCGImage:outputImage fromRect:CGRectMake(0, 0, inputImage.size.width, inputImage.size.height)];
-            });
-            
-            if (outputImageRef) {
-                outputUIImage = [UIImage imageWithCGImage:outputImageRef];
-                CGImageRelease(outputImageRef);
+        vImage_Buffer *inputBuffer;
+        vImage_Buffer *outputBuffer;
+        
+        vImage_CGImageFormat format = {
+            .bitsPerComponent = 8,
+            .bitsPerPixel = 32,
```

Comment:
Not that I can find, the box convolution functions operate on ARGB interleaved or planar 8 formatted images. We might be able to separate out the three channels to 3 separate planar 8 images and operate on them separately, but I'm not sure if that would result in the same image (I assume it would as my understanding of the box convolution just averages the surrounding pixels in each channel separately) or would have the same performance (though maybe you could calculate all three on separate threads).

But! In looking at this, I realize we don't need two image contexts and can get rid of one of them. So at least we have that going for us :)